### PR TITLE
Use storage dead statement to free up registers.

### PR DESCRIPTION
**Pull request by ptersilie**

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -108,6 +108,12 @@ impl TraceCompiler {
         }
     }
 
+    fn free_register(&mut self, local: &Local) {
+        if let Some(reg) = self.assigned_regs.get(local) {
+            self.available_regs.push(*reg);
```
Comment (vext01) [inline ykcompile/src/lib.rs:113]: What about the hashmap storing register assignments? Does the entry there need to be freed too?

I guess it depends if a local index can be reused. I don't think they can...

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -108,6 +108,12 @@ impl TraceCompiler {
         }
     }
 
+    fn free_register(&mut self, local: &Local) {
+        if let Some(reg) = self.assigned_regs.get(local) {
```
Comment (vext01) [inline ykcompile/src/lib.rs:112]: Is this guard necessary?

**Comment by vext01** [comment]
Couple of comments.

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs]: Yes, I noticed that only after I assigned you. Fix is already pushed.

Comment (vext01) [reply ykcompile/src/lib.rs]: I'm suggesting that it may not be strictly necessary. Read again.

Comment (ptersilie) [reply ykcompile/src/lib.rs]: They can be reused, e.g. by entering a function call (all arguments are assigned to $1..$N). So we should definitely clean up the `assigned_regs` too.

Though now that I think about it, entering a function call messes up `assigned_regs` already, i.e. we would have to reset and reload `assigned_regs` when entering/leaving a call, which we currently don't do. Well, investigating proper register allocation is next on my todo list anyways, so we should fix this properly.

Comment (vext01) [reply ykcompile/src/lib.rs]: Agreed. The problem you point out above will go away with a proper allocator.

File: ykcompile/src/lib.rs [hunk 3]
```diff
@@ -344,6 +344,7 @@ impl TraceCompiler {
         }
 
         tc.ret();
+        tc.crash_dump(CompileError::OutOfRegisters);
```
Comment (vext01) [inline ykcompile/src/lib.rs:347]: did you mean to commit this?

Comment (ptersilie) [reply ykcompile/src/lib.rs:347]: No, sorry.

Comment (ptersilie) [reply ykcompile/src/lib.rs]: Fixed.

**Comment by vext01** [comment]
Please squash.

File: ykcompile/src/lib.rs [hunk 4]
```diff
@@ -252,6 +258,8 @@ impl TraceCompiler {
             }
             Statement::Return => self.c_return()?,
             Statement::Call(op, args, dest) => self.c_call(op, args, dest)?,
+            Statement::StorageLive(_) => {},
+            Statement::StorageDead(l) => self.free_register(l),
```
Comment (bjorn3) [inline ykcompile/src/lib.rs:262]: What about

```rust
_1: u8;

bb0: {
    StorageLive(_1);
    _1 = 42u8;
    Goto bb2;
}

bb1: {
    StorageDead(_1);
    return;
}

bb2: {
    Use(_1);
    Goto ebb1;
}

Comment (vext01) [reply ykcompile/src/lib.rs:262]: As I understand, the input to the trace compiler is a trace, so there are no blocks to speak of. Everything should have been inlined.

The above would be:
```
    StorageLive(_1);
    _1 = 42u8;
    Use(_1);
    Goto ebb1;
    StorageDead(_1);
    return;
```

Is that right @ptersilie ?

Comment (bjorn3) [reply ykcompile/src/lib.rs:262]: /me forgot that this is a trace compiler and not a normal compiler.

Comment (vext01) [reply ykcompile/src/lib.rs:262]: No worries! You got me thinking for a moment there!

**Comment by ptersilie** [comment]
Squashed. Though it requires a 3-way merge so we better merge your PR first.

**Comment by ptersilie** [comment]
@vext01 The ykrustc build is running. Can I rebase this while we wait?

**Comment by vext01** [comment]
yep!

**Comment by ptersilie** [comment]
ykrustc has finished building so we should be able to bors this now. Let's give it a try.

bors try

**Comment by ptersilie** [comment]
Formatting again. Why do I never learn. :sob: 

bors try

**Comment by ptersilie** [comment]
@vext01 I need to squash and then this is ready for merging.

**Comment by vext01** [comment]
please squash.

**Comment by ptersilie** [comment]
Squashed.

**Pull request merged**